include/unittest.h

```diff
@@ -522,15 +546,80 @@ class unittest
         MockRepoInstanceHolder<0>::instance = 0;
     }
 
-    // -------------------------------------------------------------------------
-    // Private Interface
-    // -------------------------------------------------------------------------
+    template <typename F>
+    void expect_exception(F &&f, std::shared_ptr<const std::exception> expected)
+    {
+        bool caught = false;
+        bool wrong_exception = false;
+        std::string caught_str;
+        std::string expecting_str;
+
+        try
+        {
+            f();
+        }
+        catch (BaseException &be)
+        {
+            throw;
+        }
+        catch (bfn::general_exception &ge)
+        {
+            caught = true;
+            if (typeid(ge) != typeid(*expected))
+            {
+                wrong_exception = true;
+                caught_str = typeid(ge).name();
+                expecting_str = typeid(*expected).name();
+            }
+            else
+                std::cout << "caught: " << ge << '\n';
+        }
+        catch (std::exception &e)
+        {
+            caught = true;
+            if (typeid(e) != typeid(*expected))
+            {
+                wrong_exception = true;
+                caught_str = typeid(e).name();
+                expecting_str = typeid(*expected).name();
+            }
+            else
+                std::cout << "caught: " << e << '\n';
+        }
+        catch (...)
+        {
+            caught = true;
+            wrong_exception = true;
+            std::cerr << "unknown exception caught" << '\n';
+        }
+
+        print_wrong_exception_line(caught, wrong_exception);
```

Comment:
no. Lets prove that it works with just the VMCS code to start, and then right after provide a PR that just changes EXPECT, and I would include the expect_true, expect_false with that PR as well.